### PR TITLE
[SPARK-34257][SQL] Improve performance for last_value over unbounded window frame

### PR DESCRIPTION
**Pull request by beliefer**
### What changes were proposed in this pull request?
The current implement of `last_value` over unbounded window frame will execute `updateExpressions` multiple times.
In fact, `last_value` only execute `updateExpressions` once on the last row.

### Why are the changes needed?
Improve performance for last_value over unbounded window frame

### Does this PR introduce _any_ user-facing change?
'No'.

### How was this patch tested?
Jenkins test.

### CODE REVIEW

**Comment by AngersZhuuuu** [comment]
Do you have do some benchmark? I think with a benchmark result is good.

**Comment by cloud-fan** [comment]
ok to test

**Comment by beliefer** [comment]
cc @cloud-fan

File: sql/core/src/main/scala/org/apache/spark/sql/execution/window/WindowFunctionFrame.scala [hunk 1]
```diff
@@ -500,8 +501,18 @@ final class UnboundedWindowFunctionFrame(
     if (processor != null) {
       processor.initialize(rows.length)
       val iterator = rows.generateIterator()
-      while (iterator.hasNext) {
-        processor.update(iterator.next())
+      if (jumpToEnd) {
+        var lastRow = EmptyRow
+        while (iterator.hasNext) {
+          lastRow = iterator.next()
```
Comment (srowen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/window/WindowFunctionFrame.scala:507]: Is there any better way to get the last element in this case than iterate over all of them? maybe not given the unusual nature of the dat backing ExternalAppendOnlyUnsafeRowArray

Comment (beliefer) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/window/WindowFunctionFrame.scala:507]: `ExternalAppendOnlyUnsafeRowArray` cannot get the last element directly.

**Comment by cloud-fan** [comment]
It's a bit of waste to optimize for a single case `last_value`. Do we have other functions and get the n-th file from the end?

**Comment by beliefer** [comment]
I can't find any other fuctions like `last_value` now. Yes, it's a bit waste.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/window/WindowExecBase.scala [hunk 2]
```diff
@@ -237,6 +240,11 @@ trait WindowExecBase extends UnaryExecNode {
               new UnboundedWindowFunctionFrame(target, processor)
             }
 
+          case ("AGGREGATE_LAST", _, UnboundedPreceding, UnboundedFollowing, _) =>
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/window/WindowExecBase.scala:243]: This means that we can't use one frame to evaluate `last_value` and other aggregate functions.

TBH I'm not sure if it's worth extending the framework for such a small optimization.

Comment (beliefer) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/window/WindowExecBase.scala:243]: You mean I understand. It is not worthwhile to optimize the `last_value` alone in the framework.
I also feel that way. If we can find other functions that are the same as `last_value` in the future, then we can do it.

**Pull request closed without merging**